README.md
### Simple telegram bot to notify about EMA signal

- TAAPI https://taapi.io/indicators/exponential-moving-average/

### TODO
- [ ] Refactor code using Monad transformer
- [ ] Article or video to explain about
    - [ ] Items in the order
    ```
        > Function composition
        > Monoid 
        > Functor 
        > Applicative 
        > Monad 
        > Why we said monad can't compose 
        > Monad Transformer
    ```
    - [ ] Monad transformer

### Export env variables

```
# taapi key
export TA_API_KEY="paste your ta api key here"

# telegram chanel/group id
export CHAT_ID="paste your telegram chat group/chanel here"

# bot_key in format botxyz:AAAAA
export TELEGRAM_BOT_KEY="paste your telegram bot key here"

```

### Run haskell binary file
```
# cabal run simple-telegram-bot \
    $SYMBOL \
    $INTERVAL \
    $BACKTRACKS \
    $THRESHOLD_PERCENTAGE

cabal run simple-telegram-bot "BTC/USDT" "1h" 24 1

```

### Run bash script
link to [ema.sh](./scripts/ema.sh) file
```
# need to export env variables as above step
./scripts/ema.sh "BTC/USDT" "1h" 12 2
```


CHANGELOG.md
# Revision history for basic-http-client

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


scripts/ema.sh
#!/usr/bin/env bash

# usage:
#
# export TA_API_KEY="your taapi key"
# export CHAT_ID="your chat id"
# export TELEGRAM_BOT_KEY="your telegram bot key"
# ./scripts/ema.sh "AHA/USDT" "1h" 12 2

SYMBOL=$1
INTERVAL=$2
BACKTRACKS=$3
THRESHOLD_PERCENTAGE=$4

INDICATOR="ema"
ENDPOINT="https://api.taapi.io/$INDICATOR"
URL="$ENDPOINT?secret=$TA_API_KEY&exchange=binance&symbol=$SYMBOL&interval=$INTERVAL&backtracks=$BACKTRACKS"
RESPONSE=`curl -s $URL | jq`
FIRST_BACKTRACK=`echo $RESPONSE | jq '.[0].value|tonumber'`
LAST_BACKTRACK=`echo $RESPONSE | jq '.[-1].value|tonumber'`

GAP=`echo "$FIRST_BACKTRACK-$LAST_BACKTRACK" | bc -l`
echo "first backtrack $FIRST_BACKTRACK"
echo "last backtrack $LAST_BACKTRACK"
echo "gap $GAP"

TREND="up"
BASE=$FIRST_BACKTRACK
if (( $(echo "$FIRST_BACKTRACK > $LAST_BACKTRACK" | bc -l) )); then
  TREND="down"
  BASE=LAST_BACKTRACK
fi
ABS_GAP=${GAP#-}
PERCENTAGE=`echo "$ABS_GAP/$BASE*100" | bc -l`

echo "trend $TREND"
echo "percentage $PERCENTAGE"
echo "base $BASE"

if (( $(echo "$PERCENTAGE >= $THRESHOLD_PERCENTAGE" | bc -l) )); then
    MSG="symbol: $SYMBOL trend: $TREND percentage: $PERCENTAGE base: $BASE gap: $GAP"
    TELEGRAM_API_URL="https://api.telegram.org/$TELEGRAM_BOT_KEY/sendmessage"
    curl -s $TELEGRAM_API_URL \
        --data-urlencode "chat_id=$CHAT_ID" \
        --data-urlencode "text=$MSG" | jq
else
    echo "not reach threshold yet!"
fi



src/EmaImproved.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE  DeriveGeneric #-}

module EmaImproved where

import           Data.Aeson
                    ( Value
                    , encode
                    , Array
                    , FromJSON
                    , ToJSON
                    , Object
                    )
import qualified Data.ByteString.Char8 as S8
import qualified Data.ByteString.Lazy.Char8 as L8
import qualified Data.Yaml             as Yaml
import           Network.HTTP.Simple
import           Control.Exception
                    ( try
                    , Exception
                    , SomeException
                    )
import           Network.HTTP.Client
                    (HttpExceptionContent(StatusCodeException))
import           Data.Typeable
import           Data.Aeson.Types (parseEither)
import           Text.Read.Lex (Number)
import           GHC.Generics
import           System.Environment (getEnv)

import Control.Monad.Trans.Except
import Control.Monad.Trans.Class


data MyException =
    MyException String
        deriving (Show, Typeable)
instance Exception MyException

data Candle =
    Candle {
          backtrack :: Int
        , value :: Float
    } deriving (Show, Generic)

type Candles = [Candle]

instance FromJSON Candle
instance ToJSON Candle

getEma ::
       String
    -> String
    -> Int
    -> ExceptT MyException IO Candles
getEma symbol interval backtracks = do
    key <-  lift $ getEnv "TA_API_KEY"
    ExceptT $ do
        let params =
                [ ("secret", Just $ S8.pack key)
                , ("exchange", Just "binance")
                , ("symbol", Just $ S8.pack symbol)
                , ("interval", Just $ S8.pack interval)
                , ("backtracks", Just $ S8.pack $ show backtracks)
                ]
            request =
                  setRequestPath "/ema"
                $ setRequestHost "api.taapi.io"
                $ setRequestMethod "GET"
                $ setRequestQueryString params
                $ setRequestSecure True
                $ setRequestPort 443
                $ defaultRequest
        response <- try $ httpJSON request
        case response of
            Left err -> return $ Left err
            Right r ->  return $
                if statusCode == 200
                    then Right $ getResponseBody r
                    else Left $ MyException $ L8.unpack $ encode rBody
                where
                    statusCode = getResponseStatusCode r
                    rBody = getResponseBody r :: Candles

data TelegramMessage = TelegramMessage
    {
          symbol :: String
        , trend :: String
        , percentage :: Float
        , base :: Float
        , gap :: Float
        , firstCandleValue :: Float
        , lastCandleValue :: Float
    } deriving(Show)

handleEma ::
    String
    -> Float
    -> Candles
    -> ExceptT MyException IO String
handleEma symbol threshold cdls = ExceptT $ do
    if percentageValue >= threshold
        then return $ Right $ show msg
        else return $ Left $ MyException $ "Not reach threshold yet" ++ "; percentageValue " ++ show percentageValue ++ " | threshold "++ show threshold
    where
        firstCandleValue = value $ head cdls
        lastCandleValue  = value $ last cdls
        gap = firstCandleValue - lastCandleValue
        baseValue
            | gap > 0 = lastCandleValue
            | otherwise = firstCandleValue
        percentageValue = abs gap/baseValue * 100
        msg
            | gap > 0 = TelegramMessage
                { symbol=symbol
                , trend="down"
                , percentage=percentageValue
                , base=baseValue
                , gap=abs gap
                , firstCandleValue=firstCandleValue
                , lastCandleValue=lastCandleValue
                }
            | otherwise = TelegramMessage
                { symbol=symbol
                , trend="up"
                , percentage=percentageValue
                , base=baseValue
                , gap=abs gap
                , firstCandleValue=firstCandleValue
                , lastCandleValue=lastCandleValue
                }

data TelegramResponseData= TelegramResponseData
    {
          ok :: Bool
        , result :: Object
    } deriving(Show, Generic)

instance FromJSON TelegramResponseData
instance ToJSON TelegramResponseData

sendToTelegram ::
       String
    -> ExceptT MyException IO TelegramResponseData
sendToTelegram msg = do
    telegramKey <- lift $ getEnv "TELEGRAM_BOT_KEY"
    telegramChatID <- lift $ getEnv "CHAT_ID"
    let params =
            [ ("chat_id", Just $ S8.pack telegramChatID)
            , ("text", Just $ S8.pack msg)
            ]
        requestPath = S8.pack $ "/" ++ telegramKey ++ "/sendmessage"
        request =
              setRequestMethod "GET"
            $ setRequestHost "api.telegram.org"
            $ setRequestPath requestPath
            $ setRequestQueryString params
            $ setRequestSecure True
            $ setRequestPort 443
            $ defaultRequest
    ExceptT $ do
        response <- try $ httpJSON request
        case response of
            Left err -> return $ Left err
            Right r ->  return $
                if statusCode == 200
                    then Right $ getResponseBody r
                    else Left $ MyException $ L8.unpack $ encode rBody
                where
                    statusCode = getResponseStatusCode r
                    rBody = getResponseBody r :: TelegramResponseData

justDoIt ::
       String
    -> String
    -> Int
    -> Float
    -> ExceptT MyException IO String
justDoIt symbol interval backtracks threshold = do
    candles <- getEma symbol interval backtracks
    emaSignal <- handleEma symbol threshold candles
    teleMsg <- sendToTelegram emaSignal
    return $ show teleMsg



src/Ema.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE  DeriveGeneric #-}

module Ema where

import           Data.Aeson
                    ( Value
                    , encode
                    , Array
                    , FromJSON
                    , ToJSON
                    , Object
                    )
import qualified Data.ByteString.Char8 as S8
import qualified Data.ByteString.Lazy.Char8 as L8
import qualified Data.Yaml             as Yaml
import           Network.HTTP.Simple
import           Control.Exception
                    ( try
                    , Exception
                    , SomeException
                    )
import           Network.HTTP.Client
                    (HttpExceptionContent(StatusCodeException))
import           Data.Typeable
import           Data.Aeson.Types (parseEither)
import           Text.Read.Lex (Number)
import           GHC.Generics
import System.Environment (getEnv)


data MyException =
    MyException String
        deriving (Show, Typeable)
instance Exception MyException

data Candle =
    Candle {
          backtrack :: Int
        , value :: Float
    } deriving (Show, Generic)

type Candles = [Candle]

instance FromJSON Candle
instance ToJSON Candle

getEma ::
       String
    -> String
    -> Int
    -> IO (Either MyException Candles)
getEma symbol interval backtracks = do
    secret <- try $ getEnv "TA_API_KEY"
    case secret of
        Left e -> return $ Left e
        Right key -> do
            let params =
                    [ ("secret", Just $ S8.pack key)
                    , ("exchange", Just "binance")
                    , ("symbol", Just $ S8.pack symbol)
                    , ("interval", Just $ S8.pack interval)
                    , ("backtracks", Just $ S8.pack $ show backtracks)
                    ]

                request =
                      setRequestPath "/ema"
                    $ setRequestHost "api.taapi.io"
                    $ setRequestMethod "GET"
                    $ setRequestQueryString params
                    $ setRequestSecure True
                    $ setRequestPort 443
                    $ defaultRequest

            response <- try $ httpJSON request
            case response of
                Left err -> return $ Left err
                Right r ->  return $
                    if statusCode == 200
                        then Right $ getResponseBody r
                        else Left $ MyException $ L8.unpack $ encode rBody
                    where
                        statusCode = getResponseStatusCode r
                        rBody = getResponseBody r :: Candles

data TelegramMessage = TelegramMessage
    {
          symbol :: String
        , trend :: String
        , percentage :: Float
        , base :: Float
        , gap :: Float
        , firstCandleValue :: Float
        , lastCandleValue :: Float
    } deriving(Show)


handleEma :: String -> Float -> Candles -> IO (Either MyException String)
handleEma symbol threshold cdls = do
    if percentageValue >= threshold
        then return $ Right $ show msg
        else return $ Left $ MyException $ "Not reach threshold yet" ++ "; percentageValue " ++ show percentageValue ++ " | threshold "++ show threshold
    where
        firstCandleValue = value $ head cdls
        lastCandleValue  = value $ last cdls
        gap = firstCandleValue - lastCandleValue
        baseValue
            | gap > 0 = lastCandleValue
            | otherwise = firstCandleValue
        percentageValue = abs gap/baseValue * 100
        msg
            | gap > 0 = TelegramMessage
                { symbol=symbol
                , trend="down"
                , percentage=percentageValue
                , base=baseValue
                , gap=abs gap
                , firstCandleValue=firstCandleValue
                , lastCandleValue=lastCandleValue
                }
            | otherwise = TelegramMessage
                { symbol=symbol
                , trend="up"
                , percentage=percentageValue
                , base=baseValue
                , gap=abs gap
                , firstCandleValue=firstCandleValue
                , lastCandleValue=lastCandleValue
                }


data TelegramResponseData= TelegramResponseData
    {
          ok :: Bool
        , result :: Object
    } deriving(Show, Generic)

instance FromJSON TelegramResponseData
instance ToJSON TelegramResponseData


sendToTelegram ::
       String
    -> IO (Either MyException TelegramResponseData)
sendToTelegram msg = do
    keyEnv <- try $ getEnv "TELEGRAM_BOT_KEY"
    case keyEnv of
        Left e -> return $ Left e
        Right key -> do
            chatIDEnv <- try $ getEnv "CHAT_ID"
            case chatIDEnv of
                Left e -> return $ Left e
                Right chatID -> do
                    let params =
                            [ ("chat_id", Just $ S8.pack chatID)
                            , ("text", Just $ S8.pack msg)
                            ]
                        requestPath = S8.pack $ "/" ++ key ++ "/sendmessage"
                        request =
                              setRequestMethod "GET"
                            $ setRequestHost "api.telegram.org"
                            $ setRequestPath requestPath
                            $ setRequestQueryString params
                            $ setRequestSecure True
                            $ setRequestPort 443
                            $ defaultRequest

                    response <- try $ httpJSON request

                    case response of
                        Left err -> return $ Left err
                        Right r ->  return $
                            if statusCode == 200
                                then Right $ getResponseBody r
                                else Left $ MyException $ L8.unpack $ encode rBody
                            where
                                statusCode = getResponseStatusCode r
                                rBody = getResponseBody r :: TelegramResponseData

